### ISSUE TITLE
Nested codes

### ISSUE THREAD

**Issue by elinw** [ropenscilabs/qcoder#16]
Currently nested codes in a document don't work.

**Comment by elinw** [comment]
Also probably this solution should include indicating an invalid tag set, specifically an open without a close or a close without an open. That should actually run on save?

**Comment by learithe** [comment]
I've written a function that can handle nested qcodes now, it's a significant change since you can't just do it with a straight regex in R (as far as I know you can't do recursive regex in R?). I'll fit my solution into into the existing code and put it up soon. 

For flagging an invalid tag set ... with nested codes I don't think we could necessarily tell which open or close went with which for reporting the specific invalid tag? We could however do the following pretty easily: 1) indicate if the number of (QCODE) and (/QCODE) tags do not match, 2) indicate if there is a (/QCODE) not immediately followed by `{#code}`.  Would that be sufficient?  How would you like this to be reported?

**Comment by elinw** [comment]
Okay I’ve been looking at this too, but that’s great.  Invalid I would say should you’ve identified the key ones, not sure it if is possible to mismatch other than that.  
Regex doesn’t really work well for nesting, from my understanding it is a parsing problem, which also is more standard for unstructured text.  I was looking at the stringr str_locate_all http://stringr.tidyverse.org/reference/str_locate.html <http://stringr.tidyverse.org/reference/str_locate.html> function which gives you the start and end points of the strings ..is that what you are using?   Excited to see what you have come up with!

What would be good I think is an error.  We’ll have to think about this because right now we are using this when pulling the quotes back out, not when saving the coding.  I’m wondering if in the long run we’ll want to validate the tags on save (separate function)
?  That may be more of a wish list item.

**Comment by learithe** [comment]
I've submitted a pull request. I also added some simple error checking and generated warnings if these are detected. 

I don't see an easy way with my function to identify what specific text block is missing an opening or closing block. I can think about it a bit -- if you spot one, add it! It would be a lot easier if we had a tree structure, which I suppose could be easier to build with your str_locate_all solution?
